Separate ref & decomp registries

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #80731
* #80497

Separate ref and decomp registries so that refs (which usually decompose
to prims) and decomps (which are usually aten-to-aten decompositions)
can be separated.

davidberard98 committed:
Separate ref & decomp registries

Separate ref and decomp registries so that refs (which usually decompose
to prims) and decomps (which are usually aten-to-aten decompositions)
can be separated.

[ghstack-poisoned]

davidberard98 committed:
Update on "Separate ref & decomp registries"


Separate ref and decomp registries so that refs (which usually decompose
to prims) and decomps (which are usually aten-to-aten decompositions)
can be separated.

[ghstack-poisoned]

SherlockNoMad commented:
@ezyang and @mruberry 

It's useful to have 3 tables: aten2aten_decomp, aten2prim_decomp, decomposition_table. 
decomposition_table is the union of aten2aten_decomp and aten2prim_decomp.

davidberard98 committed:
Update on "Separate ref & decomp registries"


Separate ref and decomp registries so that refs (which usually decompose
to prims) and decomps (which are usually aten-to-aten decompositions)
can be separated.

[ghstack-poisoned]

davidberard98 committed:
Update on "Separate ref & decomp registries"


Separate ref and decomp registries so that refs (which usually decompose
to prims) and decomps (which are usually aten-to-aten decompositions)
can be separated.

[ghstack-poisoned]

davidberard98 committed:
Update on "Separate ref & decomp registries"


Separate ref and decomp registries so that refs (which usually decompose
to prims) and decomps (which are usually aten-to-aten decompositions)
can be separated.

[ghstack-poisoned]

davidberard98 committed:
Update on "Separate ref & decomp registries"


Separate ref and decomp registries so that refs (which usually decompose
to prims) and decomps (which are usually aten-to-aten decompositions)
can be separated.

[ghstack-poisoned]

davidberard98 committed:
Update on "Separate ref & decomp registries"


Separate ref and decomp registries so that refs (which usually decompose
to prims) and decomps (which are usually aten-to-aten decompositions)
can be separated.

[ghstack-poisoned]

ezyang commented on torch/_refs/__init__.py:
```diff
@@ -310,7 +310,22 @@ def __maybe_broadcast(x, shape):
-from torch._decomp import register_decomposition
+from torch._decomp import register_decomposition, decomposition_table
+
+ref_decomposition_table: Dict[torch._ops.OpOverload, Callable] = {}
```
Needs documentation

ezyang commented:
Given the experience with the previous PR, I fear that people are extremely unlikely to get the distinction here correctly without some testing. Unfortunately, the testing here is not as straightforward; maybe running things in the aten2aten set and checking if they actually call into prims.

On the PR itself, I'm a little ambivalent about the change here. This PR codifies a clear separation between aten and prim and this generally makes sense and it is not too difficult for people to know how to do it. So there is not much cost. But when I talk to most of our eager backend customers (e.g. executorch), it seems clear that they are looking for an operator set which is going to be some hybrid of prim operators and aten operators, since they DO rely on getting fused operators for their backends and can't assume a magic compiler will work it all out. So not only is "I only want prims" or "I only want aten ops" not what they are asking for, but they sometimes need even more fine grained distinctions within, say, aten ops (I want some composite implicit autograd ops to be preserved verbatim). So even though distinguishing between aten and prim may be nice and simple for us, it doesn't seem to be enough for the real users, and we should design accordingly.

davidberard98 committed:
Update on "Separate ref & decomp registries"


Separate ref and decomp registries so that refs (which usually decompose
to prims) and decomps (which are usually aten-to-aten decompositions)
can be separated.

[ghstack-poisoned]

davidberard98 committed:
Update on "Separate ref & decomp registries"


Separate ref and decomp registries so that refs (which usually decompose
to prims) and decomps (which are usually aten-to-aten decompositions)
can be separated.

[ghstack-poisoned]
